### PR TITLE
add viridian (Virus assembler from amplicon sequencing reads)

### PR DESCRIPTION
**Pull request by dguibert**
The PR adds viridian, viridian-workflow and their dependencies.

It works on x86_64 and aarch64.

### CODE REVIEW

**Comment by alalazo** [comment]
@spackbot fix style

**Review by alalazo — DISMISSED** [review]
Can you update license years at the top of the files?

**Comment by dguibert** [comment]
> Can you update license years at the top of the files?

Yes, I'll do.

**Comment by dguibert** [comment]
> > Can you update license years at the top of the files?
> 
> Yes, I'll do.

Hi @alalazo, done :smile:

File: var/spack/repos/builtin/packages/py-varifier/package.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    """Variant call verification"""
+
+    homepage = "https://github.com/iqbal-lab-org/varifier"
+    url = "https://github.com/iqbal-lab-org/varifier/archive/refs/tags/v0.3.1.tar.gz"
+
+    version(
+        "master",
+        git="https://github.com/iqbal-lab-org/varifier",
+        branch="master",
+        numeric_equivalent=">",
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-varifier/package.py]: What is this argument? Why is that needed?

Comment (dguibert) [reply var/spack/repos/builtin/packages/py-varifier/package.py]: good question... They are not needed.

**Comment by alecbcs** [comment]
ping @alalazo

**Comment by adamjstewart** [comment]
Needs to be rebased, minimap2 has already been updated with newer versions.

File: var/spack/repos/builtin/packages/minimap2/package.py [hunk 2]
```diff
@@ -32,3 +52,9 @@ def install_minimap2(self):
         make(*make_arg)
         mkdirp(prefix.bin)
         install("minimap2", prefix.bin)
+        sed = which("sed")
+        if self.spec.satisfies("js_engine=node"):
+            sed("-ie", "1s/k8/node/", "./misc/paftools.js")
+            # sed("-ie", "1s/k8/node/", "./misc/mmphase.js")
+        install("./misc/paftools.js", prefix.bin)
+        # install("./misc/mmphase.js", prefix.bin)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/minimap2/package.py]: What are these things that are commented out? Can they be removed?

We shouldn't use sed as it won't exist on Windows, we should use `filter_file` instead.

File: var/spack/repos/builtin/packages/py-cluster-vcf-records/package.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+from spack.package import *
+
+
+class PyClusterVcfRecords(PythonPackage):
+    """Python package to cluster VCF records. Used by gramtools and minos"""
+
+    homepage = "https://github.com/iqbal-lab-org/cluster_vcf_records"
+
+    url = "https://github.com/iqbal-lab-org/cluster_vcf_records/archive/refs/tags/v0.13.2.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cluster-vcf-records/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-cluster-vcf-records/package.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    version("0.13.1", sha256="5cc8768a8fd28fe20052641d6d83cc7599d709608c7b99297f8ef64547266049")
+    version("0.13.0", sha256="f86d87d4941b5c8409bcb653557c9a52086d8e20fd9cc295bd7785a37fc67c6c")
+    version("0.12.4", sha256="785942d5dd2ec5c969a9a3af9704717ed5f9354e7e47682d7c89a84999e652f1")
+    version("0.12.3", sha256="ab42c5bce13c192a12613383c5fc705e81f21c01050d9c5872fcd9774488af9d")
+    version("0.12.2", sha256="f16b643e6a52df88e9d149571fb2560acb0dcf05641cc6d9090ab22285e8713f")
+    version("0.12.1", sha256="8b8dfd6a38c29abe22fe7d0fc9a36ac8c57b8ab4782b8d4afb3e1f8a3470400d")
+    version("0.12.0", sha256="d3aaf7779f77edef89bd701a73ff9c31525ba60a0807539d64300172bdad2919")
+    version("0.11.1", sha256="02d8686150af76a7cb635c7b0149e85381f25b79d225a7052c10f3e9c4c1d601")
+    version("0.11.0", sha256="81bb3892f6383d9d82c6bd6a135f263a5c534034f931ef73693d6b434fa0ca6a")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cluster-vcf-records/package.py:27]: Missing build dep on setuptools

File: var/spack/repos/builtin/packages/py-fastaq/package.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyFastaq(PythonPackage):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fastaq/package.py]: Correct name is py-pyfastaq: https://pypi.org/project/pyfastaq/

File: var/spack/repos/builtin/packages/py-pymummer/package.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    version("0.10.3", sha256="cabae6ff65cdafc0d52c3a824c0e8456423f740260c6f91e1853554a6b13d42f")
+    version("0.10.2", sha256="3c0c7308b88fcef7e43589992e31d1384165095acee253680b5844f9e2feb7b1")
+    version("0.10.1", sha256="8c358227b842a61477889db50af0918ea3e793fddc56f6a9900e6820f2712b20")
+    version("0.10.0", sha256="381927a3dfee34727da173bfc747b34a9ddecb1647c2d36c3df4ceb297f85309")
+    version("0.9.0", sha256="e0c8bead3dacd1ecb29cdb1bdd60d7b52f13073dfbb152eefcfe08a147b9b3a1")
+    version("0.8.1", sha256="b7b137ac1e96fdaa24a18a56dc35db26c276b70615b511303518d3771631e189")
+    version("0.8.0", sha256="45d33f9aaa92a1ad06732456d3b8539cf09a81aac5bbfb3f61dd571b85c74116")
+    version("0.7.1", sha256="33ab7aeb4f09a4e00ca32c237d3f6eca0050127663ad445b139debf7137a4e57")
+    version("0.7.0", sha256="bd329946f67d3c7ebb783038077cbd831a63d65353e24f9454404150e5217876")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pymummer/package.py:26]: missing build dep on setuptools

File: var/spack/repos/builtin/packages/py-qcovid/package.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class PyQcovid(PythonPackage):
+    """QC pipelines for sars-cov-2 sequence+consensus submitted to the ENA"""
+
+    homepage = "https://github.com/iqbal-lab-org/QCovid"
+    url = "https://github.com/iqbal-lab-org/QCovid"
+
+    version("master", git="https://github.com/iqbal-lab-org/QCovid", branch="master")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qcovid/package.py:17]: Missing build dep on setuptools. Dep types are prob wrong, I assume these are needed at build/run.

File: var/spack/repos/builtin/packages/py-varifier/package.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    homepage = "https://github.com/iqbal-lab-org/varifier"
+    url = "https://github.com/iqbal-lab-org/varifier/archive/refs/tags/v0.3.1.tar.gz"
+
+    version("master", git="https://github.com/iqbal-lab-org/varifier", branch="master")
+    version("0.3.1", sha256="119f49d985f694ef0452dd70d2c53134b09a82a95d356913f7d729617c8cc7c1")
+    version("0.3.0", sha256="d318db2ef96eebc9682db9e0b11178f9ebee891afb584ff6aa32de3728d56f36")
+    version("0.2.0", sha256="9ac7ad3eacd21261ca7a5b4a214152cb55e51cfb8c0cb3ffa3538ebfebc8e359")
+    version("0.1.0", sha256="1f58fbb5d38f8427df0521eda5e92d916283e14ffffc0008ff49cc9fcf009707")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-varifier/package.py:20]: Missing build dep on setuptools

File: var/spack/repos/builtin/packages/py-varifier/package.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    depends_on("bcftools", type="run")
+    depends_on("vt", type="run")
+    depends_on("py-biopython", type=("build", "run"))
+    depends_on("py-cluster-vcf-records@0.13.2:", type=("build", "run"))
+    # mappy >= 2.17 provided by minimap2
+    depends_on("minimap2@2.17:", type=("build", "run"))
+    depends_on("py-pandas")
+    depends_on("py-fastaq@3.14.0:")
+    depends_on("py-pymummer")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-varifier/package.py]: buid/run?

File: var/spack/repos/builtin/packages/py-varifier/package.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    depends_on("py-pysam", type="run")
+    depends_on("py-seaborn", type=("build", "run"))
+
+    @run_before("install")
+    def configure(self):
+        sed = which("sed")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-varifier/package.py]: filter_file

File: var/spack/repos/builtin/packages/py-viridian-workflow/package.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+
+class PyViridianWorkflow(PythonPackage):
+    """Viridian Workflow"""
+
+    homepage = "https://github.com/iqbal-lab-org/viridian_workflow"
+
+    url = "https://github.com/iqbal-lab-org/viridian_workflow/archive/refs/tags/v0.3.1.tar.gz"
+
+    version("master", git="https://github.com/iqbal-lab-org/viridian_workflow", branch="master")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-viridian-workflow/package.py]: `git` can be moved to the package-level like `url`

File: var/spack/repos/builtin/packages/py-viridian-workflow/package.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on("py-pysam", type=("build", "run"))
+
+    depends_on("bcftools", type="run")
+    depends_on("vt", type="run")
+    depends_on("minimap2", type="run")
+    depends_on("racon", type="run")
+    depends_on("py-viridian", type="run")
+    depends_on("mummer", type="run")
+    depends_on("py-varifier", type="run")
+    depends_on("py-qcovid", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-viridian-workflow/package.py]: Where are these documented?

File: var/spack/repos/builtin/packages/py-viridian/package.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyViridian(PythonPackage):
+    """Virus assembler from amplicon sequencing reads."""
+
+    homepage = "https://github.com/iqbal-lab-org/viridian"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-viridian/package.py]: This repo has been renamed to cylon, but the cylon and viridian names have already been taken on PyPI, so we need to call it something else. It also says not to use this package, use viridian-workflow instead. Do we need this package?

File: var/spack/repos/builtin/packages/py-viridian/package.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    version("main", git="https://github.com/iqbal-lab-org/viridian", branch="main")
+    version("0.1.0", sha256="b319a4dc6d6510a9bd4cb6248c38d38f2d11c02eed7eb5645fe48d683c3c1e9f")
+
+    # depends_on('python@2.X:2.Y,3.Z:', type=('build', 'run'))
+    depends_on("py-setuptools", type=("build", "run"))
+    # depends_on('py-foo',        type=('build', 'run'))
+    depends_on("racon", type="run")
+    depends_on("minimap2", type=("build", "run"))
+    depends_on("py-pysam")
+    depends_on("py-fastaq")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-viridian/package.py]: Missing deps, wrong types, remove comments

File: var/spack/repos/builtin/packages/racon/package.py [hunk 15]
```diff
@@ -11,9 +11,19 @@ class Racon(CMakePackage):
     """Ultrafast consensus module for raw de novo genome assembly of long
     uncorrected reads."""
 
-    homepage = "https://github.com/isovic/racon"
+    homepage = "https://github.com/lbcb-sci/racon"
     url = "https://github.com/isovic/racon/releases/download/1.2.1/racon-v1.2.1.tar.gz"
 
+    version("1.5.0", sha256="41e362f71cc03b934f17d6e2c0d626e1b2997258261b14551586de006666424a")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/racon/package.py:17]: These versions don't seem to exist on github...

File: var/spack/repos/builtin/packages/racon/package.py [hunk 16]
```diff
@@ -30,7 +40,7 @@ class Racon(CMakePackage):
     conflicts("%gcc@:4.7,10.1.0:")
     conflicts("%clang@:3.1")
 
-    patch("aarch64.patch", when="target=aarch64:")
+    patch("aarch64.patch", when="@:1.4.99 target=aarch64:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/racon/package.py]:
```suggestion
    patch("aarch64.patch", when="@:1.4 target=aarch64:")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Most of these comments apply to multiple packages, we should always prefer PyPI source code over GitHub tarballs.

File: var/spack/repos/builtin/packages/py-viridian-workflow/package.py (hunk 12 above)
Comment (dguibert) [reply var/spack/repos/builtin/packages/py-viridian-workflow/package.py]: from https://github.com/iqbal-lab-org/viridian_workflow/blob/master/.ci/install_dependencies.sh

File: var/spack/repos/builtin/packages/py-viridian/package.py (hunk 13 above)
Comment (dguibert) [reply var/spack/repos/builtin/packages/py-viridian/package.py]: How can cylon be named?

Yes cylon is used inside viridan-workflow
https://github.com/iqbal-lab-org/viridian_workflow/blob/94466d79560d1480cc003bb56e782bebfe31032c/viridian/cylon.py#L18

File: var/spack/repos/builtin/packages/racon/package.py (hunk 15 above)
Comment (dguibert) [reply var/spack/repos/builtin/packages/racon/package.py:17]: racon has been moved to a new repo. The url has been updated.

File: var/spack/repos/builtin/packages/py-viridian/package.py (hunk 13 above)
Comment (dguibert) [reply var/spack/repos/builtin/packages/py-viridian/package.py]: the cylon package on pypi is hosted on https://github.com/RocknRoot/Cylon, the last version had been released on 2013.
The project is archived since Jan 14, 2021 and deprecated by its author.

**Comment by dguibert** [comment]
Hi @adamjstewart 
I've tried to apply all of your comments, much appreciated, thanks.

**Comment by dguibert** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-viridian/package.py (hunk 13 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-viridian/package.py]: See https://github.com/pypi/warehouse/issues/4004#issuecomment-621980052 for the process to claim the "cylon" name on PyPI.

File: var/spack/repos/builtin/packages/py-cylon/package.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    git = "https://github.com/iqbal-lab-org/cylon"
+
+    version("main", branch="main")
+    version("0.1.0", sha256="4c61c5b600df9c64fb2c323b48642f185ff2cdc8b7b52f06c9827683899501eb")
+
+    depends_on("py-setuptools", type=("build", "run"))
+    depends_on("racon", type="run")
+    depends_on("minimap2", type=("build", "run"))
+    depends_on("py-pysam")
+    depends_on("py-pyfastaq")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cylon/package.py:24]: type=build/run

File: var/spack/repos/builtin/packages/py-cylon/package.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyCylon(PythonPackage):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cylon/package.py:9]: Naming discussion ongoing.

File: var/spack/repos/builtin/packages/py-qcovid/package.py [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    homepage = "https://github.com/iqbal-lab-org/QCovid"
+    url = "https://github.com/iqbal-lab-org/QCovid"
+    git = "https://github.com/iqbal-lab-org/QCovid"
+
+    version("master", branch="master")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-pysam")
+    depends_on("minimap2")
+    depends_on("py-sqlalchemy@1.3.20:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qcovid/package.py:21]:
```suggestion
    depends_on("py-sqlalchemy@1.3.20")
```
Shouldn't these 3 be build/run deps?

File: var/spack/repos/builtin/packages/py-varifier/package.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+    depends_on("bcftools", type="run")
+    depends_on("vt", type="run")
+    depends_on("py-biopython", type=("build", "run"))
+    depends_on("py-cluster-vcf-records@0.13.2:", type=("build", "run"))
+    # mappy >= 2.17 provided by minimap2
+    depends_on("minimap2@2.17:", type=("build", "run"))
+    depends_on("py-pandas")
+    depends_on("py-pyfastaq@3.14.0:")
+    depends_on("py-pymummer")
+    depends_on("py-pysam", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-varifier/package.py:31]: Shouldn't these all be type=build/run?

File: var/spack/repos/builtin/packages/py-viridian-workflow/package.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    version("0.2.2", sha256="3f83d2080c739fe3f5a8d6d21e9c43324a1c7bc4b75d7817341dc8a31cf53679")
+    version("0.2.1", sha256="0c3225d2485f39070faa7bee4eecd0573efdd1b114313db3b5d4b2126038676c")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-cluster-vcf-records@0.13.3:", type=("build", "run"))
+    depends_on("py-matplotlib", type=("build", "run"))
+    depends_on("py-intervaltree", type=("build", "run"))
+    depends_on("py-pyfastaq", type=("build", "run"))
+    depends_on("py-pysam", type=("build", "run"))
+    depends_on("py-requests@2.26.0:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-viridian-workflow/package.py:30]: At least in 0.3.2, cluster-vcf-records, matplotlib, and requests aren't even imported or mentioned anywhere in the code.

File: var/spack/repos/builtin/packages/racon/package.py [hunk 22]
```diff
@@ -11,26 +11,28 @@ class Racon(CMakePackage):
     """Ultrafast consensus module for raw de novo genome assembly of long
     uncorrected reads."""
 
-    homepage = "https://github.com/isovic/racon"
-    url = "https://github.com/isovic/racon/releases/download/1.2.1/racon-v1.2.1.tar.gz"
-
-    version("1.4.3", sha256="dfce0bae8234c414ef72b690247701b4299e39a2593bcda548a7a864f51de7f2")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/racon/package.py:17]: Would be preferable to first deprecate these old versions and remove them after the next minor release so that this doesn't come as a surprise to any users. You can add a `url_for_version` that chooses the appropriate url to use.

File: var/spack/repos/builtin/packages/racon/package.py [hunk 23]
```diff
@@ -11,26 +11,28 @@ class Racon(CMakePackage):
...
+    version("1.4.20", sha256="50c22e66a1141c865b70917eecf34155b9ab3a4642765381c30f64557bdb2a9f")
+    version("1.4.13", sha256="e91e6a7170554230089ebea19b2c2b5134a1677247830d8c5a3b6aa686241387")
+    version("1.4.12", sha256="658e04c6f282f53bcc8ffd2c7f97a787b97447d41fef29727eddd96d576efbae")
+    version("1.4.11", sha256="f1a45b5a8928ed2e133e6262a0887a99ec19dfba9143cdb8d4747d4530fa699e")
+    version("1.4.10", sha256="101c1ba967eda296fea6f88c7ee7b01eb5c912b253dd0b0fdd80399b504808cb")
+    version("1.4.9", sha256="ae0bef208271166dd6325440131e408386408c12a9497d8d743504908ecb45e8")
+    version("1.4.7", sha256="75e39395ced636dd0f5ef1abe431d8f9f6404eb0a806c530cff8e4fa4f57fab7")
+    version("1.4.6", sha256="6514e89e4e6eea1d14fc1b760c4ccd02a833443284259d52628eeea0eb5f70db")
 
     depends_on("cmake@3.2:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/racon/package.py:28]: cmake 3.11+ required in latest version, not sure when that changed

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by alecbcs** [comment]
@dguibert did you get a chance to take a look at the review feedback. Is this PR still something you'd like to pursue?

**Comment by dguibert** [comment]
> @dguibert did you get a chance to take a look at the review feedback. Is this PR still something you'd like to pursue?

I'm waiting for the developpers to discuss and decide on https://github.com/iqbal-lab-org/cylon/issues/21.
I could apply the comments if this can reach develop afterwards by letting the naming as it is now and if we will change py-cyclon pypi one upstream made their choice.

I was a user of py-workflow-viridian, not use it anymore but wanted to share my recipes.

**Comment by adamjstewart** [comment]
If you want, you can separate the non-cylon things into a separate PR and we can merge those more quickly. We don't have to wait for cyclon to claim a spot on PyPI, just need to choose a name that isn't already on PyPI or replace the old name.

**Comment by alalazo** [comment]
Any news on this PR @dguibert ?

**Comment by dguibert** [comment]
> Any news on this PR @dguibert ?

Sorry I was waiting for the developper decision(s).
I'm going to follow @adamjstewart suggestion: separate the non-cylon things into a separate PR.

**Comment by dguibert** [comment]
- https://github.com/spack/spack/pull/42410
- https://github.com/spack/spack/pull/42411
- https://github.com/spack/spack/pull/42412
- https://github.com/spack/spack/pull/42413

Other packages will be left on https://github.com/dguibert/spack/tree/dg/iqbal-lab-pkgs

**Comment by dguibert** [comment]
split in separate PRs
